dhis2/dhis2-core
# feat: aggregate data exchange job [DHIS2-13552]

### Summary
Adds a job to run one or more aggregate data exchanges with a CRON schedule.

Job progress tracking was added to the exchange processing itself.
Each source of an exchange is processed as an independent item in a stage.
If an item fails next item is processed.
That means if multiple exchanges are added to a job each becomes its own stage/step in the job process.

The import summaries of each exchange are aggregated into a single `ImportSummaries` and added to the notifier so the result becomes available in the API.

### Manual Testing
* create one internal and one external aggregate data exchange
* create either one job with both exchanges as parameter or two jobs, each having one of the exchanges as parameter
* run the job(s) manually
* check `/api/system/taskSummaries/AGGREGATE_DATA_EXCHANGE` for summary results
* check `/api/scheduling/completed/AGGREGATE_DATA_EXCHANGE` for job tracking
* check that rerunning the job updates the summary and clears the one before
* check that a broken exchange configuration (e.g. target URL points to non existing server) results in error process result
* create a job with exchange ID that does not exist (via API directly) and check that running the job results in clean error report (does not "crash")

### Example Summary
```json
{
  "katyBT4VDKl":{
    "responseType":"ImportSummaries",
    "status":"SUCCESS",
    "imported":0,
    "updated":21,
    "deleted":0,
    "ignored":0,
    "importSummaries":[
      {
        "responseType":"ImportSummary",
        "status":"SUCCESS",
        "importOptions":{
          "idSchemes":{
            
          },
          "dryRun":false,
          "async":false,
          "importStrategy":"CREATE_AND_UPDATE",
          "mergeMode":"REPLACE",
          "reportMode":"FULL",
          "skipExistingCheck":false,
          "sharing":false,
          "skipNotifications":false,
          "skipAudit":false,
          "datasetAllowsPeriods":false,
          "strictPeriods":false,
          "strictDataElements":false,
          "strictCategoryOptionCombos":false,
          "strictAttributeOptionCombos":false,
          "strictOrganisationUnits":false,
          "requireCategoryOptionCombo":false,
          "requireAttributeOptionCombo":false,
          "skipPatternValidation":false,
          "ignoreEmptyCollection":false,
          "force":false,
          "firstRowIsHeader":true,
          "skipLastUpdated":false,
          "mergeDataValues":false,
          "skipCache":false
        },
        "description":"Import process completed successfully",
        "importCount":{
          "imported":0,
          "updated":21,
          "ignored":0,
          "deleted":0
        },
        "conflicts":[
          
        ],
        "dataSetComplete":"false"
      }
    ],
    "total":21
  }
}
```

### Example Progress Tracking
```json
[
  {
    "summary":"Aggregate data exchange completed with status SUCCESS",
    "status":"SUCCESS",
    "completedTime":"2022-08-11T10:48:13.403",
    "startedTime":"2022-08-11T10:48:13.022",
    "description":"Aggregate data exchange of 1 exchange(s)",
    "stages":[
      {
        "summary":"Aggregate data exchange completed (1/1): 'IwA3TtzogCe', type: 'INTERNAL'",
        "status":"SUCCESS",
        "completedTime":"2022-08-11T10:48:13.399",
        "startedTime":"2022-08-11T10:48:13.036",
        "description":"exchange aggregate data Internal data exchange to internal target (local)",
        "totalItems":0,
        "onFailure":"SKIP_ITEM",
        "items":[
          {
            "summary":"Imported 0, updated 21, deleted 0, ignored 0",
            "status":"SUCCESS",
            "completedTime":"2022-08-11T10:48:13.399",
            "startedTime":"2022-08-11T10:48:13.036",
            "description":"dx: fbfJHSPpUQD,cYeuwXTCPkU,Jtf34kNZhzP; pe: LAST_12_MONTHS,202201; ou: ImspTQPwCqd",
            "onFailure":"SKIP_ITEM",
            "duration":363,
            "complete":true
          }
        ],
        "duration":363,
        "complete":true
      }
    ],
    "jobId":"katyBT4VDKl",
    "duration":381,
    "complete":true
  }
]
```
